### PR TITLE
Add search modifiers to utils

### PR DESCRIPTION
**Pull request by jasonLaster**
Associated Issue: #1637 

### Summary of Changes

This PR extracts the source-search utils from #1644 and adds some unit tests around `buildQuery`

### Test Plan

- [ ] test searching
- [ ] test selecting text
- [ ] test iterating with cmd+g

### CODE REVIEW

File: src/utils/build-query.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+ * Ignore doing outline matches for less than 3 whitespaces
+ *
+ * @memberof utils/source-search
+ * @static
+ */
+function ignoreWhiteSpace(str: string): string {
+  return /^\s{0,2}$/.test(str) ? "(?!\s*.*)" : str;
+}
+
+function wholeMatch(query: string, wholeWord: boolean): string {
```
Comment (wldcordeiro) [inline src/utils/build-query.js]: Slightly simpler?

```js
function wholeMatch(query: string, wholeWord: boolean): string {
  if (query == "" || !wholeWord) {
    return query;
  }
  
  return `\\b${query}\\b`;
}

File: src/utils/source-search.js [hunk 2]
```diff
@@ -20,28 +17,19 @@ function SearchState() {
...
-/**
- * Ignore doing outline matches for less than 3 whitespaces
- *
- * @memberof utils/source-search
- * @static
- */
-function ignoreWhiteSpace(str) {
-  return /^\s{0,2}$/.test(str) ? "(?!\s*.*)" : str;
+  const regexQuery = buildQuery(query, modifiers, { isGlobal: false });
+  return cm.getSearchCursor(regexQuery, pos, !caseSensitive);
```
Comment (wldcordeiro) [inline src/utils/source-search.js]: The case sensitive flag is only useful with `cm.getSearchCursor` when the query is a string, if it's a regex then it's an ignored argument.

**Comment by wldcordeiro** [comment]
Going to pull this locally and run the tests, plus maybe push some changes. Looks great though @jasonLaster the search modifiers are a complicated set of features hah

**Comment by jasonLaster** [comment]
@wldcordeiro sounds good! Thanks :)

File: src/components/Editor/SearchBar.js [hunk 3]
```diff
@@ -78,7 +83,7 @@ const SearchBar = React.createClass({
     const changedFiles = selectedSource != prevProps.selectedSource
                           && hasLoaded;
 
-    if (doneLoading || changedFiles) {
+    if (doneLoading || changedFiles || this.state.query != prevState.query) {
```
Comment (wldcordeiro) [inline src/components/Editor/SearchBar.js]: The counts weren't updating properly before.

**Comment by jasonLaster** [comment]
Thanks @wldcordeiro!

**Pull request merged**